_drafts/zero-to-main-libc.md

```diff
@@ -0,0 +1,726 @@
+multiple processes safely.
+
+For the functions that cannot be easily made re-entrant, newlib depends on the
+operating system correctly setting the `_impure_ptr` variable whenever a context
+switch occur. That variable is expected to hold a `struct _reent` for the
+current thread. That struct is used to store state for standard library
+functions being used by that thread.
+
+#### Locking shared memory
+
+Some standard library functions depend on global memory which would not make
+sense to hold in the `_reent` struct. This is especially important when using
+`malloc` to allocate memory out of the heap. If mutliple threads try modifying
+the heap at the same time, they risk corrupting it.
+
+To allow multiple threads to call `malloc`, Newlib provides the `__malloc_lock`
+and `__malloc_unlock` APIs[^1]. A good implementation of these APIs would lock
+and unlock a recursive mutex.
+
+## Implementing our own C standard library
+
+In some cases, you may want to take different tradeoffs than the ones taken by
+the implementers of Newlib. Perhaps you are willing to sacrifice some
+functionality for code space, or are willing to trade performance for security.
+In most cases it is easier to replace a few functions, though you may end up
+with a fully custom C library.
+
+### Replacing a function
+
+Because Newlib is a static library with a separate object file for every
+function, all you need to do to replace a function is define it in your program.
+The linker won’t go looking for it in static libraries if it finds it in your
+code.
+
+For example, we may want to replace Newlib’s `printf` implementation, either
+because it is too large or because it depends on dynamic memory management.
+Using Marco Paland’s [excellent alternative](https://github.com/mpaland/printf)
+is as simple as a Makefile change.
+
+We first clone it in our firmware’s folder under `lib/printf`, and update our
+Makefile to reflect the change: 
+
+```makefile
+PROJECT := with-libc
+BUILD_DIR ?= build
+
+INCLUDES = \
+  ... \
+  lib/printf
+
+SRCS = \
+  ... \
+  lib/printf/printf.c \
+	startup_samd21.c \
+	$(PROJECT).c
+
+include ../common-standalone.mk
+```
+
+### Full replacement
+
+In some cases, you may want to do away with Newlib altogether. Perhaps you don’t
+want any dynamic memory allocation, in which case you could use  [Embedded
+Artistry’s solid alternative](https://github.com/embeddedartistry/libc). Another
+good reason to replace the version of Newlib provided by your toolchain is to
+use your own build of it because you would like to use different compile-time
+flags.
+
+Once we have copied the static lib (`.a`) for our selected libc, we disable
+Newlib with `-nostdlib` and explicitly link in our substitute library. You can
+find the resulting Makefile below:
+
+```makefile
+PROJECT := with-libc
+BUILD_DIR ?= build
+
+CFLAGS += -nostdlib
+LDFLAGS += -L../lib/embeddedartistry_libc/libc.a -lc
```

Comment:
nit: I don't think the libc at the end of the -L is needed ...

```
  -L DIRECTORY, --library-path DIRECTORY
                              Add DIRECTORY to library search path
```

Reply:
👍